bftengine/src/bftengine/DbCheckpointManager.cpp

```diff
@@ -0,0 +1,235 @@
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "DbCheckpointManager.hpp"
+#include <future>
+#include <cmath>
+#include <assertUtils.hpp>
+
+namespace bftEngine::impl {
+using Clock = std::chrono::steady_clock;
+struct HumanReadable {
+  std::uintmax_t size{};
+
+  template <typename Os>
+  friend Os& operator<<(Os& os, HumanReadable hr) {
+    int i{};
+    double mantissa = hr.size;
+    for (; mantissa >= 1024.; ++i) {
+      mantissa /= 1024.;
+    }
+    mantissa = std::ceil(mantissa * 10.) / 10.;
+    os << mantissa << "BKMGTPE"[i];
+    return i == 0 ? os : os << "B (" << hr.size << ')';
+  }
+};
+Status DbCheckpointManager::createDbCheckpoint(const CheckpointId& checkPointId,
+                                               const uint64_t& lastBlockId,
+                                               const uint64_t& seqNum) {
+  if (!maxNumOfCheckpoints_) return Status::OK();
+  if (seqNum <= lastCheckpointSeqNum_) return Status::OK();
+  ConcordAssert(dbClient_.get() != nullptr);
+  ConcordAssert(ps_.get() != nullptr);
+  if (dbCheckptMetadata_.dbCheckPoints_.find(checkPointId) == dbCheckptMetadata_.dbCheckPoints_.end()) {
+    auto start = Clock::now();
+    auto status = dbClient_->createCheckpoint(checkPointId);
+    if (!status.isOK()) {
+      LOG_ERROR(getLogger(), "Failed to create rocksdb checkpoint: " << KVLOG(checkPointId));
+      return status;
+    }
+    auto end = Clock::now();
+    lastCheckpointCreationTime_ =
+        std::chrono::duration_cast<std::chrono::seconds>(SystemClock::now().time_since_epoch());
+    auto duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+    // update metrics
+    lastDbCheckpointBlockId_.Get().Set(lastBlockId);
+    numOfDbCheckpointsCreated_++;
+    auto maxSoFar = maxDbCheckpointCreationTimeMsec_.Get().Get();
+    maxDbCheckpointCreationTimeMsec_.Get().Set(std::max(maxSoFar, static_cast<uint64_t>(duration_ms.count())));
+    auto count = numOfDbCheckpointsCreated_.Get().Get();
+    auto averageSoFar =
+        (dbCheckpointCreationAverageTimeMsec_.Get().Get() * (count - 1) + static_cast<uint64_t>(duration_ms.count())) /
+        count;
+    dbCheckpointCreationAverageTimeMsec_.Get().Set(averageSoFar);
+    metrics_.UpdateAggregator();
+    LOG_INFO(getLogger(), "rocksdb checkpoint created: " << KVLOG(checkPointId, duration_ms.count()));
+
+    lastCheckpointSeqNum_ = seqNum;
+    dbCheckptMetadata_.dbCheckPoints_.insert(
+        {checkPointId, {checkPointId, lastCheckpointCreationTime_, lastBlockId, lastCheckpointSeqNum_}});
+    while (dbCheckptMetadata_.dbCheckPoints_.size() > maxNumOfCheckpoints_) {
```

Comment:
Isn't it enough to remove only the first one?

Reply:
Yes you are right in most of the cases its enough to remove the first one. 
But, we can have some edge case like, we change configuration for numOfDbCheckpoints. Reduce it from N to N-K, then we may have to remove K db checkpoints